### PR TITLE
Feature #94173: Scroll when inserting suggestion

### LINKED ISSUE #94173: Virtually prevent pushing code down when new lines are automatically added
I would like an option to virtually prevent vscode from pushing down the current focused line when new lines are automatically added. For example, when I click to automatically add an import, vscode adds the import line and the current focused line is scrolled down. This causes a distraction because the line I'm working on is moved away even tho I didn't want to scroll.

Here is a gif showing how IntelliJ behavior when new lines are automatically added above the current selected line. Note that after the import statement is added the current line is virtually kept the same (the line number do change, of course, but the scroll is set to a new position in order to respect the old position):

intellij

The next gif show how vscode pushes down the code and loses the previous virtual position:

vscode

Would be possible to add an option to vscode behavior like IntelliJ in this situation?

### PR DESCRIPTION
**Pull request by shskwmt**
This PR fixes https://github.com/microsoft/vscode/issues/94173

### CODE REVIEW

File: src/vs/editor/contrib/suggest/suggestController.ts [hunk 1]
```diff
@@ -273,6 +273,9 @@ export class SuggestController implements IEditorContribution {
 		// keep item in memory
 		this._memoryService.memorize(model, this.editor.getPosition(), item);
 
+		// keep line number for scrolling
+		const initialLineNumber = this.editor.getPosition().lineNumber;
```
Comment (jrieken) [inline src/vs/editor/contrib/suggest/suggestController.ts:277]: You should be using this `StableEditorScrollState`

**Review by jrieken — CHANGES_REQUESTED** [review]

File: src/vs/editor/contrib/suggest/suggestController.ts (hunk 1 above)
Comment (shskwmt) [reply src/vs/editor/contrib/suggest/suggestController.ts]: @jrieken 
Thanks for your review. I've update the PR.

**Comment by jrieken** [comment]
Adding @alexdima to review the changes in the scroll state. Tested this, it kinda works. When inserting import-completion by current line doesn't move anymore, but when undoing it seems to shift up and that's not nice

**Comment by shskwmt** [comment]
@jrieken 
Thank you.
I added `pushUndoStop()` to stop undoing.

File: src/vs/editor/browser/core/editorState.ts [hunk 2]
```diff
@@ -162,4 +163,18 @@ export class StableEditorScrollState {
...
+
+	public restoreRelativeVerticalPositionOfCursor(editor: ICodeEditor, shouldPushUndoStop: boolean = false): void {
+		const currentCursorPosition = editor.getPosition();
+
+		if (!this._cursorPosition || !currentCursorPosition) {
+			return;
+		}
+
+		const offset = editor.getTopForLineNumber(currentCursorPosition.lineNumber) - editor.getTopForLineNumber(this._cursorPosition.lineNumber);
+		if (shouldPushUndoStop) {
```
Comment (alexdima) [inline src/vs/editor/browser/core/editorState.ts:175]: I don't think `pushUndoStop` belongs here. This class should not do that.

The undo stop does not record scroll positions, so any caller of `restoreRelativeVerticalPositionOfCursor` can call `editor.pushUndoStop` after `editor.setScrollTop` and there will be the same effect.

**Review by alexdima — CHANGES_REQUESTED** [review]
LGTM, except the `pushUndoStop` call in `restoreRelativeVerticalPositionOfCursor`

File: src/vs/editor/browser/core/editorState.ts (hunk 2 above)
Comment (shskwmt) [reply src/vs/editor/browser/core/editorState.ts]: @alexdima 
Thank you. I fixed.

File: src/vs/editor/browser/core/editorState.ts [hunk 3]
```diff
@@ -164,17 +164,15 @@ export class StableEditorScrollState {
...
 		if (!this._cursorPosition || !currentCursorPosition) {
 			return;
 		}
 
 		const offset = editor.getTopForLineNumber(currentCursorPosition.lineNumber) - editor.getTopForLineNumber(this._cursorPosition.lineNumber);
-		if (shouldPushUndoStop) {
-			editor.pushUndoStop();
-		}
 		editor.setScrollTop(editor.getScrollTop() + offset);
+		editor.pushUndoStop();
```
Comment (alexdima) [inline src/vs/editor/browser/core/editorState.ts:176]: As I just said above, this call does not belong here. Please remove it.

Also note the caller side where an undo stop is pushed only conditionally.

**Review by alexdima — CHANGES_REQUESTED** [review]
There is still a call to `editor.pushUndoStop` in `restoreRelativeVerticalPositionOfCursor`. It needs to be removed from there, as restoring the scrolling position should not affect the undo stack in any way.

**Comment by shskwmt** [comment]
@alexdima 
Thanks.
Could you please review and let me know if there is any problem.

File: src/vs/editor/contrib/suggest/suggestController.ts [hunk 4]
```diff
@@ -290,6 +293,9 @@ export class SuggestController implements IEditorContribution {
 			adjustWhitespace: !(suggestion.insertTextRules! & CompletionItemInsertTextRule.KeepWhitespace)
 		});
 
+		scrollState.restoreRelativeVerticalPositionOfCursor(this.editor);
+		this.editor.pushUndoStop();
```
Comment (jrieken) [inline src/vs/editor/contrib/suggest/suggestController.ts:297]: Why are you adding an undo at all?

**Review by jrieken — CHANGES_REQUESTED** [review]

File: src/vs/editor/contrib/suggest/suggestController.ts (hunk 4 above)
Comment (shskwmt) [reply src/vs/editor/contrib/suggest/suggestController.ts:297]: @jrieken 
I'm sorry, I misunderstood. I'll remove it.

**Comment by jrieken** [comment]
The scroll jump after undo still happens (https://github.com/microsoft/vscode/pull/94327#issuecomment-615153399). Adding another undo-stop is not the right approach here but I also don't know what to do. @alexdima can you advice? We now control the scroll state when *doing* the suggest edits but I don't thing we have a say when *undoing* these changes.

**Comment by alexdima** [comment]
Effectively, when *doing* and inserting an import, the editor is scrolled down by 1 line (`scrollTop` = `scrollTop` + `lineHeight`)... But when undoing the scroll top is maintained constant, the lines move up because the import line is removed.

The undo does not capture the `scrollTop` of the file because in the meantime between *doing* and *undoing* the file might have been resized, the scroll position might be somewhere outside of the cursor position... Since last milestone it is possible that the file was closed and reopened in the meantime.

That is just how undo/redo works currently, and is not 100% awesome for this specific case, and we can make new issues about that if we want to, but I don't know how to tackle it, nor do I think that it should block this PR...

**Review by alexdima — APPROVED** [review]

**Pull request merged**